File: solution/207CourseSchedule/207CourseSchedule.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
  bool dfs(vector<vector<int>> &graph, vector<bool> &visited,
           vector<bool> &onStack, int i) {
    visited[i] = true;
    onStack[i] = true;
    bool loop = false;
    for (int j = 0; j < graph[i].size(); j++) {
      if (!visited[graph[i][j]]) {
        loop = loop || dfs(graph, visited, onStack, graph[i][j]);
      } else if (onStack[graph[i][j]]) {
        return true;
      }
    }
    onStack[i] = false;
    return loop;
  }

public:
  bool canFinish(int numCourses, vector<pair<int, int>> &prerequisites) {
    vector<vector<int>> graph(numCourses);
    for (auto &p : prerequisites) {
      graph[p.second].push_back(p.first);
    }
    vector<bool> visited(numCourses, false);
    vector<bool> onStack(numCourses, false);
    for (int i = 0; i < numCourses; i++) {
      if (!visited[i]) {
        bool loop = dfs(graph, visited, onStack, i);
        if (loop) {
          return false;
        }
      }
    }
    return true;
  }
};

int main() { return 0; }

File: src/207CourseSchedule/207CourseSchedule.rs
#[derive(Clone, PartialEq)]
enum VisitState {
    Unvisited,
    Visited,
    Visiting,
}

use std::collections::HashMap;
impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let mut map:HashMap<i32, Vec<i32>> = HashMap::new();
        for pair in prerequisites {
            if let Some(neighbors) = map.get_mut(&pair[0]) {
                neighbors.push(pair[1]);
            }else{
                map.insert(pair[0], vec![pair[1]]);
            }
        }
        let mut visited = vec![VisitState::Unvisited; num_courses as usize];
        for key in map.keys() {
            if visited[*key as usize] == VisitState::Unvisited {
                visited[*key as usize] = VisitState::Visited;
                if Solution::dfs(&map, &mut visited, *key) {
                    return false;
                }
            }
        }
        true
    }
    
    fn dfs(map:&HashMap<i32, Vec<i32>>, visited: &mut Vec<VisitState>, key: i32) -> bool {
        visited[key as usize] = VisitState::Visiting;
        
        if let Some(neighbors) = map.get(&key) {
            for neihbor in neighbors {
                if visited[*neihbor as usize] == VisitState::Visiting {
                    return true;
                }
                if map.get(neihbor).is_some() {
                    if Solution::dfs(map, visited, *neihbor) {
                        return true;
                    }
                }
            }
        }
        
        visited[key as usize] = VisitState::Visited;
        return false;
    }
}
